graphql-mode.el

```diff
@@ -96,12 +101,46 @@ response from the server."
            (point))))
     (buffer-substring-no-properties start end)))
 
+(defun graphql-current-operation ()
+  "get the name of the query operation"
+  (interactive)
+  (let* ((query
+         (save-excursion
+           (replace-regexp-in-string "^[ \t\n]*" "" (graphql-current-query))))
+	 (tokens
+	  (split-string query "[ \f\t\n\r\v]+"))
+	 (first (nth 0 tokens)))
+
+    (if (string-equal first "{")
+	""
+      (replace-regexp-in-string "[({:]+" "" (nth 1 tokens)))))
+ 
+(defun graphql-current-variables ()
```

Comment:
So if I understand correctly, the syntax for the variables is

```
variables
{
   ...
}
```

?  Is it JSON?

I'm not too convinced about this. If you add variables like that, then the file doesn't contain a valid query anymore.

What if we instead define `graphql-variables` as a buffer-local variable, and open a window with a json-mode buffer to allow the user to set it interactively?  In a similar way to what we do with `graphql-url`?

Reply:
Yeah, that `variables` is a json. you are right that it will make the query file contains not query feature such as query. 

However I am not sure about the best way to put `variables` inside a normal graphql `query` file, which has mutliple query, mutation, subscription. Just "append" the variable at the end of query file. since there is no graphql parser involved.

There is one big assumption with these `variables`, and it must be the last in this query file.
in other words, one can put any `queries`/`mutation`/`subscription` *ahead* the `variables`. variables must be the last in the file.

so I add function `graphql-begining-of-variables` to find out if there is any variable in these query file. that `graphql-beginning-of-variables` will essentially only let `graphql--query` get from `(point-min)` to `(graphql-beginning-of-variables)`.
It is not the cleanest way... 

I believe `graphql-variables` is nice way to go about it.  It will be also nice to be persistence.
can we add a` .graphql-variables` file or comment out the whole variables block. so that one doesn't need to rewrite all the variables every time.....

that also apply for graphql-url as well. can we add a presistence to it too? say a pseudo comment any where in the file, just like the mode comment? at the same time make graphq-url a command, so that one can change `graphql-url` pointing to  development endpoint to production endpoint on the fly.

```
# -*- graphql-mode -*- 
# graphql-url = http://localhost:8000/graphql

query q1 {
...
}

query q2 {
...
}
mutation m1 { 
...
}

# pseudo-comment inline variables at the end or save variables in file ./.graphql-variables
# variables {
# ...
# }
```

Reply:
For my use case, I had queries in `.graphql` files.  I loaded those files with a special webpack loader as a parsed query that I then passed to apollo-client.  So the variables { .. } would break it.

My proposal would be:

1. A `graphql-variables` as buffer-local variable. The value is just passed to `graphql-query`.
2. `graphql-query` would take the variables (as an alist), serialize it and send it.
3. A command `graphql-load-variables` to load a JSON file into `graphql-variables`.

I think that would cover our use cases very nicely.  Then, in a next pull request, we can:

4. Add a command `graphql-edit-variables`. This would open a `json-mode` buffer in a dedicated window to allow the user to change `graphql-variables` and invoke `C-c C-c` to set the new value. This is more comfortable that the minibuffer.

Additionally, we can improve it even more and allow the user to press `C-x C-s` in the dedicated window to update the original file that the variables were loaded from.